### PR TITLE
[MRG][FIX] Copy of the data in get_data.

### PR DESCRIPTION
**Pull request by jaeilepp**
Closes https://github.com/mne-tools/mne-python/issues/3827
I wonder if this has always been broken...

### CODE REVIEW

File: mne/epochs.py [hunk 1]
```diff
@@ -1170,7 +1170,7 @@ def get_data(self):
         data : array of shape (n_epochs, n_channels, n_times)
             A copy of the epochs data.
         """
-        return self._get_data()
+        return self._get_data().copy()
```
Comment (agramfort) [inline mne/epochs.py]: in fact it was on purpose but the docstring was not correct.

my feeling is that the automatic copy can be problematic for large epochs objects

**Comment by mmagnuski** [comment]
👍 looks good

**Comment by jaeilepp** [comment]
I hope it's not done anywhere else. I tried looking around, but at some places the calls are quite chained so it was hard to follow.

**Comment by mmagnuski** [comment]
It might be good to add a few checks for this in some of the tests that might be affected (but that would be a separate PR, I think). But I never had this problem with other plotters (only with `plot_topo_image` - but I wasn't checking throughly).

File: mne/epochs.py [hunk 2]
```diff
@@ -1168,7 +1168,7 @@ def get_data(self):
         Returns
         -------
         data : array of shape (n_epochs, n_channels, n_times)
-            A copy of the epochs data.
+            The epochs data.
```
Comment (christianbrodbeck) [inline mne/epochs.py]: "A view on" would be more explicit

**Comment by Eric89GXL** [comment]
> I hope it's not done anywhere else. I tried looking around, but at some places the calls are quite chained so it was hard to follow.

Try setting `epochs._data.flags.writeable = False` when it is first created and you will see if/when we try to change it (hopefully).

**Comment by jaeilepp** [comment]
I'll let travis do the dirty work...

File: mne/decoding/transformer.py [hunk 3]
```diff
@@ -88,7 +88,7 @@ def fit(self, epochs_data, y):
         return self
 
     def transform(self, epochs_data, y=None):
-        """Standardize data across channels.
+        """Standardize data across channels. Operates in-place.
```
Comment (agramfort) [inline mne/decoding/transformer.py]: this should not happen. transform should not have edge effects.

Comment (jaeilepp) [inline mne/decoding/transformer.py]: Yes. I wasn't sure where this ``Scaler`` is used. Should I make it copy the data?
Also the ``y`` here is not used, but there was some comment elsewhere about compatibility with scikit learn, so I guess this can be used as a callback or something.

**Comment by agramfort** [comment]
LGTM

+1 for merge if CIs are happy

File: mne/decoding/transformer.py [hunk 4]
```diff
@@ -107,7 +107,7 @@ def transform(self, epochs_data, y=None):
             raise ValueError("epochs_data should be of type ndarray (got %s)."
                              % type(epochs_data))
 
-        X = np.atleast_3d(epochs_data)
+        X = np.atleast_3d(epochs_data).copy()
```
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: you should add a `Notes` section or say in the `epochs_data` description that it will be copied, because this could be a big memory operation

File: mne/decoding/transformer.py [hunk 5]
```diff
@@ -137,13 +137,13 @@ def inverse_transform(self, epochs_data, y=None):
...
```
Comment (Eric89GXL) [inline mne/decoding/transformer.py]: same

File: mne/viz/topo.py [hunk 6]
```diff
@@ -694,7 +694,7 @@ def plot_topo_image_epochs(epochs, layout=None, sigma=0., vmin=None,
         Figure distributing one image per channel across sensor topography.
     """
     scalings = _handle_default('scalings', scalings)
-    data = epochs.get_data()
+    data = epochs.get_data().copy()
```
Comment (Eric89GXL) [inline mne/viz/topo.py]: Instead of copying here, it could be much more memory efficient to only copy when necessary. Is it possible to do the copy on an epoch-by-epoch basis before plotting?

File: mne/decoding/transformer.py [hunk 7]
```diff
@@ -137,13 +137,13 @@ def inverse_transform(self, epochs_data, y=None):
...
 
         for key, this_pick in six.iteritems(self.picks_list_):
             if self.with_mean:
-                X[:, this_pick, :] += self.ch_mean_[key]
-            if self.with_std:
                 X[:, this_pick, :] *= self.std_[key]
+            if self.with_std:
+                X[:, this_pick, :] += self.ch_mean_[key]
```
Comment (jasmainak) [inline mne/decoding/transformer.py]: I don't quite follow this change @jaeilepp

Comment (mmagnuski) [inline mne/decoding/transformer.py]: The intent was to scale first and then center (if both scaling and centering are required), but now the conditions and operations do not match (ie. conditions should have been swapped too).

Comment (mmagnuski) [inline mne/decoding/transformer.py]: (this is because it is inverse_transform and transform first centers then scales)

Comment (jasmainak) [inline mne/decoding/transformer.py]: okay so something to fix for @jaeilepp . I am surprised there was no test to check for this.

Comment (jaeilepp) [inline mne/decoding/transformer.py]: There was a test, but the operations were done in-place so it didn't catch the error.

File: mne/decoding/transformer.py (hunk 5 above)
Comment (jaeilepp) [inline mne/decoding/transformer.py]: Done.

**Comment by agramfort** [comment]
thx @jaeilepp

File: mne/decoding/transformer.py (hunk 7 above)
Comment (mmagnuski) [inline mne/decoding/transformer.py]: @jaeilepp - It seems this got merged without correction. Now self.with_mean does `X[:, this_pick, :] *= self.std_[key]` - probably there is no test for doing centering without scaling?

Comment (jaeilepp) [inline mne/decoding/transformer.py]: Oh crap this wasn't what I meant to do...
